File: ios/expo/views/SmileIDExpoView.swift
import ExpoModulesCore
import UIKit
import SwiftUI
import SmileID

/**
 * Expo-specific implementation of SmileIDView that uses native Expo events
 */
class SmileIDExpoView: ExpoView {
    // Event dispatchers for Expo
    let onSmileIDResult = EventDispatcher()
    let onSmileIDError = EventDispatcher()
    
    // Properties from SmileIDView
    var userId: String?
    var jobId: String?
    var allowAgentMode: Bool = false
    var showInstructions: Bool = true
    var skipApiSubmission: Bool = false
    var showAttribution: Bool = true
    var extraPartnerParams: [String: String] = [:]
    
    // Hosting controller for SwiftUI views
    var hostingController: UIHostingController<AnyView>?
    
    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        clipsToBounds = true
        setupView()
    }
    
    private func setupView() {
        // Initial setup
        backgroundColor = .clear
    }
    
    /**
     * Must be overridden by subclasses to render their specific content
     */
    func renderContent() {
        fatalError("renderContent must be implemented by subclass")
    }
    
    /**
     * Helper method to set content view with proper hosting controller
     */
    func setContent<Content: View>(_ content: Content) {
        // Remove existing hosting controller if any
        hostingController?.view.removeFromSuperview()
        hostingController?.removeFromParent()
        
        // Create new hosting controller
        let newHostingController = UIHostingController(rootView: AnyView(content))
        newHostingController.view.backgroundColor = .clear
        newHostingController.view.translatesAutoresizingMaskIntoConstraints = false
        
        // Add to view hierarchy
        addSubview(newHostingController.view)
        
        // Setup constraints
        NSLayoutConstraint.activate([
            newHostingController.view.topAnchor.constraint(equalTo: topAnchor),
            newHostingController.view.leadingAnchor.constraint(equalTo: leadingAnchor),
            newHostingController.view.trailingAnchor.constraint(equalTo: trailingAnchor),
            newHostingController.view.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Store reference
        hostingController = newHostingController
    }
    
    /**
     * Helper method to add a UIView with full constraints
     */
    func setUIView(_ uiView: UIView) {
        // Remove existing subviews
        subviews.forEach { $0.removeFromSuperview() }
        
        // Add the new UIView
        uiView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(uiView)
        
        // Setup constraints
        NSLayoutConstraint.activate([
            uiView.topAnchor.constraint(equalTo: topAnchor),
            uiView.leadingAnchor.constraint(equalTo: leadingAnchor),
            uiView.trailingAnchor.constraint(equalTo: trailingAnchor),
            uiView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
    }
    
    /**
     * Emit success event using Expo's EventDispatcher
     */
    func emitSuccess(_ result: Any) {
        onSmileIDResult([
            "result": result
        ])
    }
    
    /**
     * Emit failure event using Expo's EventDispatcher
     */
    func emitFailure(_ error: Error?) {
        onSmileIDError([
            "error": error?.localizedDescription ?? "Unknown error",
            "code": "SMILE_ID_ERROR"
        ])
    }
    
    /**
     * Generic handler for SmileIDSharedResult
     */
    func handleResultCallback(_ result: SmileIDSharedResult<Any>) {
        switch result {
        case .success(let data):
            emitSuccess(data)
        case .error(let code, let message, let cause):
            emitFailure(NSError(
                domain: "SmileID",
                code: Int(code) ?? -1,
                userInfo: [
                    NSLocalizedDescriptionKey: message,
                    NSUnderlyingErrorKey: cause as Any
                ]
            ))
        case .withError(let error):
            emitFailure(error)
        }
    }
}

File: package/android/src/expo/com/smileidentity/expo/views/SmileIDExpoView.kt
package com.smileidentity.expo.views

import android.content.Context
import android.view.ViewGroup
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.ComposeView
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import com.smileidentity.SmileID
import com.smileidentity.compose.theme.colorScheme
import com.smileidentity.compose.theme.typography
import com.smileidentity.shared.SmileIDSharedResult
import kotlinx.collections.immutable.ImmutableMap
import kotlinx.collections.immutable.persistentMapOf

/**
 * Expo-specific implementation of SmileIDView that uses native Expo events
 */
abstract class SmileIDExpoView(
    context: Context, 
    appContext: expo.modules.kotlin.AppContext
) : ExpoView(context, appContext) {

    // Event dispatchers for Expo
    protected val onSmileIDResult by EventDispatcher<Map<String, Any?>>()
    protected val onSmileIDError by EventDispatcher<Map<String, Any?>>()
    
    // Properties from SmileIDView
    lateinit var composeView: ComposeView
    var userId: String? = null
    var jobId: String? = null
    var allowAgentMode: Boolean? = false
    var showInstructions: Boolean = true
    var skipApiSubmission: Boolean = false
    var showAttribution: Boolean = true
    var extraPartnerParams: ImmutableMap<String, String> = persistentMapOf()
    
    init {
        layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        setUpViews()
    }
    
    private fun setUpViews() {
        composeView = ComposeView(context)
        composeView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        addView(composeView)
        renderContent()
    }
    
    abstract fun renderContent()
    
    /**
     * Helper method to set content on the ComposeView with MaterialTheme wrapper
     */
    open fun setContentWithTheme(content: @Composable () -> Unit) {
        composeView.setContent {
            MaterialTheme(colorScheme = SmileID.colorScheme, typography = SmileID.typography) {
                Surface(content = content)
            }
        }
    }

    /**
     * Implementation of runOnUiThread for Expo
     */
    open fun runOnUiThread(action: () -> Unit) {
        post(action)
    }

    /**
     * Emit success event using Expo's EventDispatcher
     */
    open fun emitSuccess(result: String) {
        onSmileIDResult(mapOf(
            "result" to result
        ))
    }

    /**
     * Emit failure event using Expo's EventDispatcher
     */
    open fun emitFailure(error: Throwable?) {
        onSmileIDError(mapOf(
            "error" to (error?.message ?: "Unknown error"),
            "code" to "SMILE_ID_ERROR"
        ))
    }
    
    /**
     * Generic handler for SmileIDSharedResult
     */
    open fun handleResultCallback(result: SmileIDSharedResult<*>) {
        when (result) {
            is SmileIDSharedResult.Success -> {
                when (val data = result.data) {
                    is String -> emitSuccess(data)
                    else -> emitSuccess(data.toString())
                }
            }
            is SmileIDSharedResult.WithError -> emitFailure(result.cause)
            is SmileIDSharedResult.Error -> emitFailure(Exception(result.message, result.cause))
        }
    }
}